File: Algorithms/19. Remove Nth Node From End of List/python/19_remove_nth_node_from_end.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        
        len = 1
        temp = head
        while temp!= None:
            len +=1
            temp=temp.next
        
        if len == 1:
            return None
        
        size = len-n
        
        if size == 1:
            return head.next
        
        temp = head
        while size != 1:
            previous = temp
            temp = temp.next
            size -= 1
        
        
        previous.next = temp.next
        return head
    

        
    
sol = Solution()

test1 = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5, None)))))
test2 = ListNode(1, ListNode(2, None))
test3 = ListNode(1, None)

sol.removeNthFromEnd(test1,2)
sol.removeNthFromEnd(test2,2)
sol.removeNthFromEnd(test3,1)

File: Algorithms/19. Remove Nth Node From End of List/java/19_remove_nth_node_from_end.java
    // Definition for singly-linked list.
class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {

    public ListNode removeNthFromEnd(ListNode head, int n) {
    ListNode start = new ListNode();
    ListNode slow = start, fast = start;
    start.next = head;
    
    for(int i=1; i<=n+1; i++)   {
        fast = fast.next;
    }

    while (fast != null) {
        slow = slow.next;
        fast= fast.next;
    }
    
    slow.next = slow.next.next;
    return start.next;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);
        
        sol.removeNthFromEnd(head, 2);
    }
}